src/group_theory/quaternion.lean

```diff
@@ -0,0 +1,281 @@
+  { simp_rw [mul_zero, pow_zero] },
+  { rw [a_one_pow, one_def],
+    congr' 1,
+    exact zmod.nat_cast_self _, }
+end
+
+@[simp] lemma xa_pow_two (i : zmod (2 * n)) : xa i ^ 2 = a n :=
+begin
+  simp [pow_two],
+end
+
+@[simp] lemma xa_pow_four (i : zmod (2 * n)) : xa i ^ 4 = 1 :=
+begin
+  simp only [pow_succ, pow_two, xa_mul_xa, xa_mul_a, add_sub_cancel, add_sub_assoc, add_sub_cancel',
+             sub_self, add_zero],
+  norm_cast,
+  rw ← two_mul,
+  simp [one_def],
+end
+
+/--
+If `0 < n`, then `xa i` has order 4.
+-/
+@[simp] lemma order_of_xa [fact (0 < n)] (i : zmod (2 *n)) : order_of (xa i) = 4 :=
+begin
+  have h : order_of (xa i) ≤ 4 := order_of_le_of_pow_eq_one (by norm_num) (xa_pow_four i),
+  interval_cases (order_of (xa i)) with hcas,
+  { -- Show that order_of (xa i) is not 0.
+    exfalso,
+    have hpos : order_of (xa i) > 0 := order_of_pos (xa i),
+    linarith },
+  { -- Show that order_of (xa i) is not 1.
+    exfalso,
+    rw order_of_eq_one_iff at hcas,
+    injection hcas },
+  { -- Show that order_of (xa i) is not 2.
+    exfalso,
+    have hcontra : (xa i) ^ 2 = 1,
+      { rw ← hcas,
+        exact pow_order_of_eq_one _ },
+    simp at hcontra,
+    injection hcontra with hcontra',
+    apply_fun zmod.val at hcontra',
+    apply_fun ( / n) at hcontra',
+    simp only [zmod.val_nat_cast, zmod.val_zero, nat.zero_div, nat.mod_mul_left_div_self,
+               nat.div_self (_inst_1)] at hcontra',
+    norm_num at hcontra' },
+  { -- Show that order_of (xa i) is not 3.
+    exfalso,
+    have : order_of (xa i) ∣ 4,
+      { apply order_of_dvd_of_pow_eq_one,
+        simp },
+    rw hcas at this,
+    norm_num at this,
+  },
+  { -- `order_of (xa i) = 4` is the only possibility left.
+    assumption
+  }
+end
+
+/--
+If `0 < n`, then `a 1` has order `2 * n`.
+-/
+@[simp] lemma order_of_a_one [fact (0 < n)] : order_of (a 1 : quaternion_group n) = 2 * n :=
+begin
+  have hnpos : fact (0 < 2 * n) :=
+    begin rw ← zero_mul 0,
+          apply mul_lt_mul',
+            { exact zero_le 2 },
+            { assumption },
+            { exact zero_le 0 },
+            { exact nat.zero_lt_succ 1 }
+    end,
+  cases lt_or_eq_of_le (nat.le_of_dvd hnpos (order_of_dvd_of_pow_eq_one (@a_one_pow_n n))) with h h,
+  { have h1 : (a 1 : quaternion_group n)^(order_of (a 1)) = 1,
+    { exact pow_order_of_eq_one _ },
+    rw a_one_pow at h1,
+    injection h1 with h2,
+    rw [← zmod.val_eq_zero, zmod.val_nat_cast, nat.mod_eq_of_lt h] at h2,
+    exact absurd h2.symm (ne_of_lt (order_of_pos _)) },
+  { exact h }
+end
+
+/--
+If `0 < n`, then `i : zmod n` has order `n / gcd n i`
+-/
+lemma order_of_a [fact (0 < n)] (i : zmod (2 * n)) :
+  order_of (a i) = (2 * n) / nat.gcd (2 * n) i.val :=
+begin
+  haveI : fact (0 < 2 * n) :=
+    begin rw ← zero_mul 0,
+          apply mul_lt_mul',
+            { exact zero_le 2 },
+            { assumption },
+            { exact zero_le 0 },
+            { exact nat.zero_lt_succ 1 }
+    end,
```

Comment:
This appears 4 times in this file - I suggest you make it a `local attribute [instance]`